launchable/test_runners/ant.py

```diff
@@ -0,0 +1,32 @@
+import os
+from typing import List
+import click
+from . import launchable
+from ..utils.file_name_pattern import jvm_test_pattern
+
+
+@click.argument('source_roots', required=True, nargs=-1)
+@launchable.subset
+def subset(client, source_roots: List[str]):
+    def file2test(f: str):
+        if jvm_test_pattern.match(f):
+            f = f[:f.rindex('.')]   # remove extension
+            # directory -> package name conversion
+            cls_name = f.replace(os.path.sep, '.')
+            return [{"type": "class", "name": cls_name}]
+        else:
+            return None
+    
+    for root in source_roots:
+        # split "src/**/*Test.java" to "src/" and "**/*Test.java"
+        glob = "**/*"
+        if '*' in root:
+          idx = root.find('*')
+          glob = root[idx:]
+          root = root[:idx]
```

Comment:
I think this is actually problematic.

The easiest case to think of would be what if the input is `src/test/java/com/launchableinc/integration_tests/**/*Test.java`
It fails to produce the correct fully qualified class names.

This is why a fileset has the "dir" part and includes/excludes part. A file set isn't just a collection of files. It's a base directory, plus a collection of files within it.

One way to solve this is to restrict `source_roots` to be base directories. Then define `--includes PATTERN` and `--excludes PATTERN`, somewhat like how `tar` does it, which has the same structure of the modeling as we do. That is, `launchable subset ant src/test/java --includes 'com/launchableinc/integration_test/**/*Test.java'`

I recognize in order to support multiple source roots, this would have to make `--includes` and `--excludes` behave positionally, which pushes the envelope of how we use Click. This also requires expanding `Optimize` class with this new way of specifying things. This will be worth making it reusable as it won't be unique to Ant, which makes me wonder how we did other Java test runners, which generally has the same problem.

If that's a little too much to take on for Ant, which is arguably a minority test runner in 2021, maybe our starting point is to limit `source_roots`  to be directories and we hard code the default GLOB mask. That gets the ball moving until this limited mode of the support becomes inadequate.

I'd love to hear if you see other ways to solve this.

Reply:
Got it. In my understanding, recent Java projects use `src/test/java/` file path but old projects use `src/` directly, am I right? If so, we need to support both in Java projects. 

Unfortunately, I don't much time today. I fix it with the easiest way you pointed out `**/*Test.java`. But I'm going to create a ticket for supporting these file paths.

Reply:
When Ant was in active use, no such directory name convention was established. People used all sorts of random directory names to house the source roots, so I certainly wouldn't treat `src/test/java` or `src` any differently.

(BTW, `src/test/java` is the convention established by Maven and I think Gradle followed that somewhat)